prover/src/sha_compress/sha_compress_stark.rs

```diff
@@ -0,0 +1,1139 @@
+        &self,
+        inputs_and_timestamps: Vec<([u8; NUM_INPUTS], usize)>,
+        min_rows: usize,
+    ) -> Vec<[F; NUM_SHA_COMPRESS_COLUMNS]> {
+        let num_rows = inputs_and_timestamps
+            .len()
+            .max(min_rows)
+            .next_power_of_two();
+
+        let mut rows = Vec::with_capacity(num_rows);
+        for input_and_timestamp in inputs_and_timestamps.iter() {
+            let row_for_compress = self.generate_trace_rows_for_compress(*input_and_timestamp);
+            rows.push(row_for_compress);
+        }
+
+        while rows.len() < num_rows {
+            rows.push([F::ZERO; NUM_SHA_COMPRESS_COLUMNS]);
+        }
+        rows
+    }
+
+    fn generate_trace_rows_for_compress(
+        &self,
+        input_and_timestamp: ([u8; NUM_INPUTS], usize),
+    ) -> [F; NUM_SHA_COMPRESS_COLUMNS] {
+        let timestamp = input_and_timestamp.1;
+        let inputs = input_and_timestamp.0;
+
+        let mut row = ShaCompressColumnsView::<F>::default();
+        row.timestamp = F::from_canonical_usize(timestamp);
+        row.is_normal_round = F::ONE;
+        // read inputs
+        row.input_state = inputs[0..256]
+            .iter()
+            .map(|x| F::from_canonical_u8(*x))
+            .collect::<Vec<F>>()
+            .try_into()
+            .unwrap();
+        row.w_i = inputs[256..288]
+            .iter()
+            .map(|x| F::from_canonical_u8(*x))
+            .collect::<Vec<F>>()
+            .try_into()
+            .unwrap();
+        row.k_i = inputs[288..320]
+            .iter()
+            .map(|x| F::from_canonical_u8(*x))
+            .collect::<Vec<F>>()
+            .try_into()
+            .unwrap();
+
+        // compute
+        row.e_rr_6 = rotate_right(row.input_state[get_input_range(4)].try_into().unwrap(), 6);
+        row.e_rr_11 = rotate_right(row.input_state[get_input_range(4)].try_into().unwrap(), 11);
+        row.e_rr_25 = rotate_right(row.input_state[get_input_range(4)].try_into().unwrap(), 25);
+        row.s_1 = xor3(row.e_rr_6, row.e_rr_11, row.e_rr_25);
+
+        row.e_and_f = and_op(
+            row.input_state[get_input_range(4)].try_into().unwrap(),
+            row.input_state[get_input_range(5)].try_into().unwrap(),
+        );
+
+        row.not_e_and_g = andn_op(
+            row.input_state[get_input_range(4)].try_into().unwrap(),
+            row.input_state[get_input_range(6)].try_into().unwrap(),
+        );
+
+        row.ch = xor_op(row.e_and_f, row.not_e_and_g);
+
+        (row.inter_1, row.carry_1) = wrapping_add(
+            row.input_state[get_input_range(7)].try_into().unwrap(),
+            row.s_1,
+        );
+
+        (row.inter_2, row.carry_2) = wrapping_add(row.inter_1, row.ch);
+
+        (row.inter_3, row.carry_3) = wrapping_add(row.inter_2, row.k_i);
+
+        (row.temp1, row.carry_4) = wrapping_add(row.inter_3, row.w_i);
+
+        row.a_rr_2 = rotate_right(row.input_state[get_input_range(0)].try_into().unwrap(), 2);
```

Comment:
Can we just keep track of  row.input_state's [0..2], [2..13], [13..22], [22..256] ?  This also means a_rr_13 is not necessary to be tracked by each row. 
so we don't need input_state to be of size 256?  this can significantly reduce the cells scale.